### ISSUE TITLE
`pcs_wait_for`: enabling/disabling resources with pcs-modules-2

### ISSUE THREAD

**Issue by OndrejHome** [OndrejHome/ansible.pcs-modules-2#28]
Hi, this topic is also vital to me, but for the most part i had gone from just creating and deleting resources just to a plain management of them.
Currently i don't try to interface with the meta attributes, because it has proven to be not reliable in terms of the end result.

I modified the the pcs_resource this way.

i added 2 new states of the resource : enable, disable
that directly tells pcs to enable or disable the resource, just via the command

pcs resource enable \ pcs resource disable

using refreshing and cleaning up before it's start.
In the end it looks something like this,
The module will try to enable or disable the resource, wait a bit (parametrized) and if it's starting or stopping will report a changed status, if anything else will just fail.
This approach proven to be more efficient in terms of the expected results.
But may be a bit out of place(in terms of ansible philosophy probably) due to the fact i made the functionality directly in pcs_resource. I think i would move the functinality some where like pcs_manage, but in terms of cmd usage, enable\disable IMHO is better than meta.

_Originally posted by @Hoijima in https://github.com/OndrejHome/ansible.pcs-modules-2/issues/27#issuecomment-485406644_

**Comment by OndrejHome** [comment]
Thank you for feedback. Now I think I understand better your intention. It is not only enabling/disabling resources but also some additional operations around them.

`pcs resource enable` and `pcs resource disable` are essentially adding or removing the `target-role: stopped' meta attribute to/from resource. That is why I mentioned to have a look at previous issue. So if only running that command is desired I believe that using 'meta' attribute should be enough.

But as you want to add timeout and also report what was the state I feel that this is doing a little bit too much :)

To me this looks like following list of steps:
1. Check if the resource is enabled/disabled and determine if it needs changing
2. If change is needed, do the enable/disable
3. Wait up to some timeout (parametrized)
4. Determine the state of resource (running/stopped or enabled/disabled)
5. Report success or failure of change

I'm not sure what step 4 would check, if the 'enable/disable' succeeded or if the resource has stopped.
Could you please let me know what you have in mind here?

If the idea is to make sure that 'enable'/'disable' task succeeds only when resource is 'started'/stopped', then I would consider maybe splitting this into smaller things - have one module that does 'enable/disable' and one module that waits for the state to reach the desired one and fails only after timeout or on error. What do you think about this approach?

(The second part to me looks a little bit like following module that "waits for something to happen" - https://docs.ansible.com/ansible/latest/modules/wait_for_module.html ) I can imagine having something like that for here that would wait for the desired state of resource for a defined (parametrized) time.

Let me know your thought on above. If you have some code available feel free to post link to it here and I can have a look to see on how you have approached this.

**Comment by Hoijima** [comment]
Well I'll try to make a Meta properties version, I'm still working on the module anyway.

>Determine the state of resource (running/stopped or enabled/disabled)
Idea is simple, try to check 3 times with some timeout between check the server state, if the state would not be the same as we need - report a Fail.
If the state changes before 3 times, just break the loop and continue processing.

>If the idea is to make sure that 'enable'/'disable' task succeeds only when resource is 'started'/stopped', then I would consider maybe splitting this into smaller things

Then lets split it to ,idn, pcs_state_manager maybe?

**Comment by OndrejHome** [comment]
Let me know how the testing of meta attributes went for you.

For naming I would prefer something like 'pcs_wait_for' if we are aiming for module that will succeed only when given resource reached the desired state or fail in case of timeout. Something that will be really similar in way how module 'wait_for' works - https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/utilities/logic/wait_for.py

For your use case I can then imagine task like below:
~~~
- pcs_wait_for:
    resource_name: some_resource
    resource_state: started
    node: node1 #this will be optional
    delay: 10 #delay first check for 10 seconds - optional
    sleep: 5 #wait 5 seconds between checks - optional
    timeout: 300 #maximum number of seconds to wait before failing - optional
~~~

The above would really make sense as module for making aure that resources are in state that we expect and I think it would also apply to your use case. Let me know if the above would suit your case and if you would like to contribute the code for this? 
Also if you have any question feel free to ask. 

PS: I would avoid the name 'pcs_state_manager' as that gives idea that it can change state while I have in mind module that just reports success once desired state is reached.

**Comment by Hoijima** [comment]
>    node: node1 #this will be optional

According to [redhat doc](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html/configuring_the_red_hat_high_availability_add-on_with_pacemaker/s1-resourceopts-haar) we don't have node definition or something similar for pcs resource meta, did you meant something else? 

>     timeout: 300 #maximum number of seconds to wait before failing - optional
Didn't quite understand the meaning, timeout for waiting pcs to return status?Or timeout waiting for the module to finish processing :)

besides that ,structure seems ok

- pcs_wait_for:
    resource_name: some_resource
    resource_state: started
    delay: 10 #delay first check for 10 seconds - optional
    sleep: 5 #wait 5 seconds between checks - optional

**Comment by OndrejHome** [comment]
> > node: node1 #this will be optional
> 
> According to [redhat doc](https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/6/html/configuring_the_red_hat_high_availability_add-on_with_pacemaker/s1-resourceopts-haar) we don't have node definition or something similar for pcs resource meta, did you meant something else?

Here I meant that we wanna wait for resource to be for example started on 'nodeX' when defined. However thinking more about this it will not fit well the Cloned resources so lets skip this one for now.

> > ```
> > timeout: 300 #maximum number of seconds to wait before failing - optional
> > ```
> 
> Didn't quite understand the meaning, timeout for waiting pcs to return status?Or timeout waiting for the module to finish processing :)

Timeout of the module. Lets keep all 'pcs' timeouts at their defaults. I have rarely seen that timeout for 'pcs' command needed to be adjusted.

---

Also thank you for the PR #29. At present I see that you use `pcs resource cleanup` in several places and `pcs resource meta`. My thought here was to use the meta attribute change that can be already used with `pcs_resource` module instead of making it into this new `pcs_wait_for` module. The `pcs_wait_for` should ideally not change any states and just exit when some state is achieved or on timeout. See the examples below.

Example A with enabling the resource.
~~~
- name: create the resource 'test' or make sure that it is enabled
  pcs_resource:
    name: 'test'
    resource_type: 'ocf:pacemaker:Dummy'

- name: wait for the resource 'test' to get started somewhere in cluster
  pcs_wait_for:
    resource_name: 'test'
    resource_state: 'started'
~~~

Example B with disabling the resource:
~~~
- name: create the resource 'test' or make sure that it is disabled
  pcs_resource:
    name: 'test'
    resource_type: 'ocf:pacemaker:Dummy'
    options: 'meta target-role=Stopped'

- name: wait for the resource 'test' to get stopped
  pcs_wait_for:
    resource_name: 'test'
    resource_state: 'stopped'
~~~

Does the above make sense to you?

**Comment by Hoijima** [comment]
Well i tried what you suggested, but it seems we will have to modify the resource role then.

when i try to do this

```
- name: start role
  pcs_resource:
    name: "{{ item }}"
    resource_type: 'ocf:devexperts:dx'
    options: 'app={{ item }} projectuser=dxtest meta target-role=Stopped'
  become: true
  with_list: "{{ apps }}"
```
i get this in the end

```
 server (ocf::devexperts:dx):   FAILED ops (disabled, blocked)
 multiplexor    (ocf::devexperts:dx):   FAILED ops (disabled, blocked)
 activator      (ocf::devexperts:dx):   FAILED ops (disabled, blocked)
 quoting        (ocf::devexperts:dx):   FAILED ops (disabled, blocked)
 dx-ipfservice  (ocf::devexperts:dx):   FAILED ops (disabled, blocked)
 muxquote-in    (ocf::devexperts:dx):   FAILED ops (disabled, blocked)
 muxquote-out   (ocf::devexperts:dx):   FAILED ops (disabled, blocked)
```
and if i try to get them started with meta, they just will get blocked.

**Comment by OndrejHome** [comment]
From outputs it looks that resources are creates as disabled - which is expected. But the FAILED state means that there is something that failed with resource itself - most probably during initial `probe`/`monitor` or `start` operation. 
Is the same happening when you use the `ocf::pacemaker:Dummy` resource? 
If not could you check if the `ocf::devexperts:dx` resource agents reacts properly to `probe` operation?

**Comment by Hoijima** [comment]
@OndrejHome sorry for the long reply, Russia was on loooong holidays : )
Currently we have settled with using my custom module i written (without using pcs_resource) and will investigate in the matter of out custom resource, it's a legacy code, so maybe the original writer made it a bit buggy.
For now i will probably cancel my pull request, some of my colleagues will continue my work on this task instead of me.
I will write you a letter when i hear about that my guy's came up with on the behalf of your suggestion : )
Cheers!

**Comment by OndrejHome** [comment]
Hi @Hoijima, Thank you for information.
In my experience I have seen mostly poorly written LSB scripts that people tried in pacemaker cluster and complained that cluster is misbehaving :) 

I will keep this Issue open in case you or your colleagues came later, but I may also look at it myself at some point. I believe that `pcs_wait_for` module as described here makes sense, but may not have not  the priority for development.

Thank you for raising up the request here and providing the code and feedback! 
Feel free to get in touch if there is any update on your side.

**Comment by OndrejHome** [comment]
Hi @Hoijima, going through older issue I just wanted to drop here information about existance of `crm_wait_for` module that in context of this issue was referenced as `pcs_wait_for` - [Ansible galaxy: crm_wait_for](https://galaxy.ansible.com/OndrejHome/crm_wait_for) / [Github repo: ansible.crm_wait_for](https://github.com/OndrejHome/ansible.crm_wait_for).

I believe that `crm_waif_for` and use of `meta` attributes should be sufficient to get functionality described here - enable/disable resource and wait for resource to transition into enabled/disabled state or fail. The waiting part for transition is customizable with `crm_wait_for` so user can choose how long to wait for transition to finish.

In spite of above I will close issue here, but if you believe that above combination doesn't address the need feel free to re-open it or create new Issue. 
Thank you!
